docs/standards/rest.md

```diff
@@ -197,6 +245,40 @@ The following parameters are reserved for specific purposes in our API. These na
 
 The `version` URL query parameter, `?version=version_string` is reserved for selecting the API version when making a request.
 
+### <a id="collection-counts-parameters"></a>Collection counts
+
+Resource collections may provide summary counts in the response top-level `meta` object when requested, using the [collection counts metadata schema](#meta-collection-counts) defined above.
+
+This parameter is only relevant for resource collection endpoints.
+
+#### `meta_count` query parameter
+
+May support the following enumerated values:
+
+- `only`: The collection response must contain collection counts in the top-level `meta` object property. Response must not contain a `data` array of resources.
+- `with`: The collection response must contain collection counts in the top-level `meta` property, as well as paginated resources in the `data` property.
+
+A resource collection may choose to support only one of these options. For example, it may only allow counts with `only`, or `with`. This may be expressed by declaring the enum with only one of these values.
+
+#### `meta_count_by` query parameter
+
+Collections may use the `meta_count_by` query parameter to include counts grouped by one or more resource attributes in the response. The parameter must be declared in OpenAPI as an array of enumerated values using the same representation as other array parameters, `{"style": "form", "explode": false}`. Each enumerated value supported must match a resource attribute property name.
+
+#### Default behavior and interactions
+
+Unless specifically requested, the default behavior when these parameters are not provided is to _not_ include counts in the response.
+
+If `meta_count` is not specified but `meta_count_by` is specified, `meta_count=only` is implied. The following table represents the default behaviors of different combinations of these parameters and their resulting affect on the response contents:
+
+| `meta_count` value | `meta_count_by` value | Response has `.meta.count` | Response has `.meta.count_by` | Response has `.data[]` |
+|------------|---------------|--------|-|-|
+| _not set_ | _not set_ | ❌ | ❌ | ✅ |
+| _not set_ | _one or more group(s)_ | ✅ | ✅  | ❌ |
+| `only` | _*_ | ✅ | _*_ | ❌ |
+| `with` | _*_ |  ✅ | _*_ | ✅ |
+
+An endpoint may choose to change this default behavior. If so, it must be documented in the parameter(s) description(s).
```

Comment:
Should this be a stronger suggestion in the standard? I'm thinking how in v1 there are some APIs that support pagination and some that don't. I almost wonder if the ideal is that some level of the resource tree should have a default meta_count functionality supported.

Reply:
Obtaining the total count and grouped counts is likely to be a separate query that has to be made, and possibly an index scan to get an exact count. For a sufficiently large collection, you might cache this, or estimate counts in a HyperLogLog structure instead (but that's another thing to then store and keep in sync). This makes me hesitant to declare it a MUST across the board for all resources up front, before we know what we're in for here.

The things we have going for us here though, that v1 lacks:

- The OpenAPI spec will still declare whether a given resource collection supports counts, and how to get it.
- Customer demand can then drive improvement in those specs which do not have counts, with new version releases.

We could decide later that API version releases after a certain date must include counts. I think that would need an ADR for the reasons above. If we have that though, and we're confident that services can handle it, I'd feel comfortable that we could roll out such a constraint for all API versions releasing after that decision date.

Reply:
It seems reasonable to make the _behavior_ standard though... "if you implement counts, this is how it works". Just wanted to separate out these different threads.

Reply:
Shored up the wording on this one. @jcsackett had similar concerns, I'm convinced.